src/wrapper.c

```diff
@@ -112,47 +112,107 @@ int cgroup_add_all_controllers(struct cgroup *cgroup)
 {
 	struct cgroup_controller *cgc;
 	struct controller_data info;
+	enum cg_setup_mode_t mode;
 	void *handle;
-	int ret;
+	int ret = 0;
 
-	/* go through the controller list */
-	ret = cgroup_get_all_controller_begin(&handle, &info);
-	if ((ret != 0) && (ret != ECGEOF)) {
-		fprintf(stderr, "cannot read controller data: %s\n", cgroup_strerror(ret));
-		return ret;
-	}
+	if (!cgroup)
+		return ECGINVAL;
+
+	mode = cgroup_setup_mode();
+
+	/*
+	 * Per kernel documentation, cgroup-v2.rst, /proc/cgroups is "meaningless" for cgroup v2.
+	 * Use the cgroup's cgroup.controllers file instead
+	 */
+	if (mode == CGROUP_MODE_UNIFIED) {
```

Comment:
Github has mangled this diff.  I added the `if (mode == CGROUP_MODE_UNIFIED)` section, and the `else` section remained unchanged other than the indentation.